flexmeasures/data/services/utils.py

```diff
@@ -0,0 +1,90 @@
+import functools
+from flask import current_app
+from rq.job import Job
+
+import hashlib
+import base64
+
+from flexmeasures.data.models.time_series import Sensor
+
+
+def make_hash_sha256(o):
+    """
+    SHA256 instead of Python's hash function because apparently, python native hashing function
+    yields different results on restarts.
+    Source: https://stackoverflow.com/a/42151923
+    """
+    hasher = hashlib.sha256()
+    hasher.update(repr(make_hashable(o)).encode())
+    return base64.b64encode(hasher.digest()).decode()
+
+
+def make_hashable(o):
+    """
+    Function to create hashes for dictionaries with nested objects
+    Source: https://stackoverflow.com/a/42151923
+    """
+    if isinstance(o, (tuple, list)):
+        return tuple((make_hashable(e) for e in o))
+
+    if isinstance(o, dict):
+        return tuple(sorted((k, make_hashable(v)) for k, v in o.items()))
+
+    if isinstance(o, (set, frozenset)):
+        return tuple(sorted(make_hashable(e) for e in o))
+
+    if isinstance(o, Sensor):
+        return tuple(
+            (
+                make_hashable(getattr(o, attr, None))
+                for attr in ["attributes", "annotations", "id"]
+            )
+        )
+
+    return o
+
+
+def hash_function_arguments(args, kwags):
+    return make_hash_sha256(args) + make_hash_sha256(kwags)  # concat two hashes
+
+
+def redis_cache(queue):
+    """
+    Decorator that checks if a function has alread been called with the same arguments and
+    fetches the job using the job id, which is stored in Redis.
+    """
+
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            # Creating a hash from args and kwargs
+            args_hash = hash_function_arguments(args, kwargs)
+
+            # Checking if there exist the key hash in the redis queue
+            if current_app.queues[queue].connection.exists(args_hash):
+                current_app.logger.info(
+                    f"The function {func.__name__} has been called alread with the same arguments. Skipping..."
+                )
+
+                # get job id
+                job_id = current_app.queues[queue].connection.get(args_hash).decode()
+
+                # get job object from job id
+                return Job.fetch(
+                    job_id, connection=current_app.queues[queue].connection
+                )
+            else:  # new call
+
+                # call function
+                job = func(*args, **kwargs)
+
+                # store function call in redis
+                current_app.queues[queue].connection.set(
+                    args_hash, job.id
+                )  # setting return value of function call to the hash of its inputs
+
+                return job  # todo: try catch, if it fails, don't hash it or impose a max retry number
```

Comment:
What might fail here: setting up the job?

Reply:
Yes, or It might be already running/done and the job might have failed.

Reply:
True true, I shouldn't leave todo's. What I've done now is to allow creating new jobs in case that the job that the previous job had failed. Please, let me know if that's what you had in mind :D

Reply:
I marked this conversation as *unresolved* in case you missed it, @victorgarcia98 and @nhoening.

Reply:
Oh, I missed it. Thanks for the notice.

I see 1 & 2 as reasonable pros and cons but I'm not sure of having the job id as an id of the requested "question". Maybe the hash should? 

It seems that the retry functionality will retry at most `max` times to run the same function call in case of failure, not on request. The logic I'm proposing would allow users to trigger the same schedule in the case they added, for example, new beliefs that were missing. 

If we want to keep the number of tries, we can do that, for instance, in a serialized [JSON](https://redis.io/docs/data-types/strings/) like this:

```
'{"job_id" : "1212asda", "retries" : 10 }'
```

Other that this, I'm thinking that this whole caching logic could we confusing or not desired in events of a user wanting to update their schedules with new TimelyBeliefs (e.g new forecast). In those cases, the user might send the same schedule params but want to actually trigger a schedule. What do you think?

Reply:
Sorry, my link to the python-rq docs was to the wrong retry functionality: [here](https://python-rq.org/docs/jobs/#requeuing-failed-jobs) is how to requeue manually (in our case, via API request).

It's a nice idea to return information on the number of retries via our API. That could be a separate issue (needs more discussion), if we decide that. It connects to the discussion of having a RESTful API for scheduling jobs.

Your last point is indeed a real concern. Basically the state of any sensor data relevant to the scheduling problem needs to be fresh. Two ways I see we could handle that:
1. Make the sensor's hash change whenever new sensor data is recorded (possibly just within the relevant time window), so we know when the sensor's state description should no longer be considered fresh.
2. Let the API user override the whole caching logic.

Reply:
I've added the two little changes in commit [f80bb9d](https://github.com/FlexMeasures/flexmeasures/pull/616/commits/f80bb9d2684684105205dc6d2373621c83f6a4e6), we can always roll it back. Let me know if that would be it for this PR :smile:

Reply:
I just read this long thread, and I'm happy with how it played out!

Reply:
Thanks! :D Me too.